client/client_test.go

```diff
@@ -0,0 +1,171 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package client
+
+import (
+	"database/sql"
+	"io/ioutil"
+	"net"
+	"strings"
+	"testing"
+
+	"github.com/google/trillian"
+	"github.com/google/trillian/extension/builtin"
+	"github.com/google/trillian/server"
+	"github.com/google/trillian/util"
+	"google.golang.org/grpc"
+)
+
+const (
+	createSQLFile = "../storage/mysql/storage.sql"
+	mysqlURI      = "root@tcp(127.0.0.1:3306)/"
+)
+
+var mysqlURIdb = "root@tcp(127.0.0.1:3306)/unittest"
+
+type Env struct {
+	grpcServer *grpc.Server
+	logServer  *server.TrillianLogRPCServer
+	Client     *LogClient
+}
+
+// Listen opens a random high numbered port for listening.
+func Listen() (string, net.Listener, error) {
+	lis, err := net.Listen("tcp", ":0")
+	if err != nil {
+		return "", nil, err
+	}
+	_, port, err := net.SplitHostPort(lis.Addr().String())
+	if err != nil {
+		return "", nil, err
+	}
+	addr := "localhost:" + port
+	return addr, lis, nil
+}
+
+// RestDB drops and recreates the test database.
+// TODO: convert to CreateDB when registry can take a db object.
+func ResetDB() error {
+	builtin.MySQLURIFlag = &mysqlURIdb
+
+	db, err := sql.Open("mysql", mysqlURI)
+	if err != nil {
+		return err
+	}
+	defer db.Close()
+
+	resetSQL := []string{
+		"DROP DATABASE IF EXISTS unittest;",
+		"CREATE DATABASE unittest;",
+		"GRANT ALL ON unittest.* TO 'unittest'@'localhost' IDENTIFIED BY 'zaphod';",
+	}
+	for _, sql := range resetSQL {
+		if _, err := db.Exec(sql); err != nil {
+			return err
+		}
+	}
+
+	// Reconnect to use the new test database.
+	dbTest, err := sql.Open("mysql", mysqlURIdb)
+	if err != nil {
+		return err
+	}
+	defer dbTest.Close()
+
+	createSQL, err := ioutil.ReadFile(createSQLFile)
+	if err != nil {
+		return err
+	}
+
+	sqlSlice := strings.Split(string(createSQL), ";\n")
+	// Iteriate from 0 to len-1. Last element of slice is "".
+	for _, sql := range sqlSlice[:len(sqlSlice)-1] {
+		if _, err := dbTest.Exec(sql); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+// NewEnv creates a fresh DB, log server, and client.
+func NewEnv() (*Env, error) {
+	err := ResetDB()
+	if err != nil {
+		return nil, err
+	}
+
+	logID := int64(1234)
+	timesource := &util.SystemTimeSource{}
+	registry, err := builtin.NewDefaultExtensionRegistry()
+	if err != nil {
+		return nil, err
+	}
+
+	grpcServer := grpc.NewServer()
+	logServer := server.NewTrillianLogRPCServer(registry, timesource)
+	trillian.RegisterTrillianLogServer(grpcServer, logServer)
+
+	// Listen and start server.
+	addr, lis, err := Listen()
+	if err != nil {
+		return nil, err
+	}
+	go grpcServer.Serve(lis)
+
+	// Connect to the server.
+	cc, err := grpc.Dial(addr, grpc.WithInsecure())
+	if err != nil {
+		return nil, err
+	}
+
+	return &Env{
+		grpcServer: grpcServer,
+		logServer:  logServer,
+		Client:     New(logID, cc),
+	}, nil
+}
+
+// Close shuts down the server.
+func (env *Env) Close() {
+	env.grpcServer.Stop()
```

Comment:
I omit dropping the database here so as to leave some evidence of what went wrong after a test.  I drop the database during test setup.